Add tests for async functions

bterlson opened the pull request:
Been sitting on these for too long. Please take a look and let me know what needs fixing up and where people would like to see more coverage.

goyakin commented:
/cc @ianwjhalliday

rwaldron commented on test/built-ins/AsyncFunction/AsyncFunction-is-subclass.js:
```diff
@@ -0,0 +1,15 @@
+async function foo() { };
+var AsyncFunction = foo.constructor;
+assert.sameValue(Object.getPrototypeOf(AsyncFunction), Function, "Prototype of constructor");
+assert.sameValue(Object.getPrototypeOf(AsyncFunction.prototype), Function.prototype, "Prototype of constructor's prototype");
```
The assertion messages in these two could be clearer, but no one should block on that.

rwaldron commented:
Since `AsyncFunction` is not exposed as a property of the global object, as `Function` is, there should be a single test that asserts this

rwaldron commented on test/built-ins/AsyncFunction/AsyncFunction-name.js:
```diff
@@ -0,0 +1,12 @@
+---*/
+
+var AsyncFunction = async function foo() { }.constructor;
+assert.sameValue(AsyncFunction.name, "AsyncFunction");
```
Needs the descriptor verification assertions

rwaldron commented on test/built-ins/AsyncFunction/instance-has-name.js:
```diff
@@ -0,0 +1,12 @@
+
+async function foo () { };
+
+assert.sameValue(foo.name, "foo");
```
Needs the descriptor verification assertions

rwaldron commented on test/language/expressions/async-arrow-function/arrow-returns-promise.js:
```diff
@@ -0,0 +1,15 @@
+  Async arrow functions return promises
+---*/
+
+var p = (async () => await 1 + await 2)();
```
Is it worthwhile to include an `instanceof` check?

jugglinmike replied:
Or even more precisely, a `Object.getPrototypeOf` check

rwaldron commented on test/language/expressions/await/await-BindingIdentifier-nested.js:
```diff
@@ -0,0 +1,15 @@
+author: Brian Terlson <brian.terlson@microsoft.com>
+es7id: pending
+description: >
+  Await is allowed as an identifier in functions nested in async functions
```
Should read: "is not"

rwaldron commented on test/language/expressions/await/await-awaits-thenable-not-callable.js:
```diff
@@ -0,0 +1,19 @@
+
+async function foo() {
+  var thenable = { then: 42 };
+  var res = await thenable
```
Semi-colon?

rwaldron commented on test/language/expressions/await/await-in-nested-generator.js:
```diff
@@ -0,0 +1,20 @@
+author: Brian Terlson <brian.terlson@microsoft.com>
+es7id: pending
+description: >
+  Await is allowed as an identifier in functions nested in async functions
```
"identifier in functions nested in async functions" => "identifier in **generator** functions nested in async functions"

rwaldron commented on test/language/expressions/object/method-definition/early-errors-object-method-NSPL-with-USD.js:
```diff
@@ -0,0 +1,13 @@
+negative: SyntaxError
+---*/
+({
+  foo(x = 1) {"use strict"}
```
Was this intended to be an ArrowFunction?

bterlson replied:
No, this is a method definition.

rwaldron commented on test/language/expressions/object/method-definition/object-method-returns-promise.js:
```diff
@@ -0,0 +1,15 @@
+author: Brian Terlson <brian.terlson@microsoft.com>
+es7id: pending
+description: >
+  Async function expressions return promises
```
"function expressions" => "method definition"

rwaldron commented on test/language/statements/async-function/declaration-returns-promise.js:
```diff
@@ -0,0 +1,13 @@
+
+async function foo() { };
+var p = foo();
+assert(p instanceof Promise, "async functions return promise instances");
```
Ah, I see this now—ignore my quesion from an earlier test

rwaldron commented:
It's possible that I simply missed this, but I was looking for a negative test that shows a failure when the AwaitExpression is missing an operand, eg:

```js
async function foo() {
  await;
}
```

jugglinmike commented on test/built-ins/AsyncFunction/instance-construct.js:
```diff
@@ -0,0 +1,16 @@
+---*/
+
+assert.throws(TypeError, function() {
+  async function foo() { }
```
Recommend moving this statement outside of the assertion body, in the unlikely event that the implementation throws something like `new TypeError('Terrible function name')`

jugglinmike commented on test/language/expressions/async-arrow-function/arrow-returns-promise.js:
```diff
@@ -0,0 +1,15 @@
+p.then(function (v) {
+  assert.sameValue(v, 3);
+  $DONE();
+}, $DONE);
```
I recently started explicitly invoking `$DONE` with a error description from within a closure, just in case the promise was rejected with out a reason (or with any falsey reason) since [the helper wouldn't report an error in those cases](https://github.com/tc39/test262/blob/fbad34959ffedb02c5b8c040e0f992a6fc006f36/harness/doneprintHandle.js#L5)

jugglinmike commented on test/language/expressions/async-function/early-errors-expression-formals-contains-super-call.js:
```diff
@@ -0,0 +1,11 @@
+negative: SyntaxError
+---*/
+
+(async function foo (foo = super()) { let bar; });
```
The `let` binding isn't necessary here

jugglinmike commented on test/language/expressions/async-function/early-errors-expression-formals-contains-super-property.js:
```diff
@@ -0,0 +1,11 @@
+negative: SyntaxError
+---*/
+
+(async function foo (foo = super.foo) { let bar; });
```
The `let` binding isn't necessary here

jugglinmike commented on test/language/expressions/await/await-awaits-thenable-not-callable.js:
```diff
@@ -0,0 +1,19 @@
+  assert.sameValue(res, thenable);
+}
+
+foo().then($DONE, $DONE);
```
This relies on Promise semantics, which are not under test. To some extent, there's no avoiding this, but we can mitigate some issues by incrementing a global `callCount` variable after the assertion, and asserting *that* prior to invoking `$DONE`.

Then again, this is ES7, so maybe we have to start taking correct Promise operation for granted...

jugglinmike commented on test/language/expressions/await/await-in-function.js:
```diff
@@ -0,0 +1,12 @@
+  Await is an identifier in a function
+---*/
+
+function foo(await) { return await }
```
Missing semi-colon for the `return` statment

jugglinmike commented on test/language/expressions/await/await-in-generator.js:
```diff
@@ -0,0 +1,12 @@
+  Await in a generator is an identifier
+---*/
+
+function* foo(await) { yield await };
```
Missing semi-colon for the `yield` ExpressionStatement

jugglinmike commented on test/language/expressions/await/await-in-nested-function.js:
```diff
@@ -0,0 +1,20 @@
+var await;
+async function foo() {
+  function bar() {
+    await = 1;
```
Nice one!

jugglinmike commented on test/language/statements/async-function/evaluation-body-that-throws-after-await.js:
```diff
@@ -0,0 +1,23 @@
+}
+
+foo().then(function () {
+  $ERROR("Should not be called");
```
You'll want to call `$DONE("Should not be called")` since [`$ERROR` simply throws an error](https://github.com/tc39/test262/blob/fbad34959ffedb02c5b8c040e0f992a6fc006f36/harness/sta.js#L16-L19), which in this context would reject the promise but not signal test failure.

jugglinmike commented on test/language/statements/async-function/evaluation-body-that-throws.js:
```diff
@@ -0,0 +1,21 @@
+}
+
+foo().then(function () {
+  $ERROR("Should not be called");
```
Same as above w/r/t `$ERROR`

jugglinmike commented on test/language/statements/async-function/evaluation-body.js:
```diff
@@ -0,0 +1,18 @@
+let called;
+async function foo() {
+  called = true;
+  await new promise();
```
I guess it doesn't matter much here, but `new promise()` will trigger a ReferenceError

jugglinmike commented on test/language/statements/async-function/evaluation-default-that-throws.js:
```diff
@@ -0,0 +1,19 @@
+---*/
+async function foo(x = null++) {}
+foo().then(function () {
+  $ERROR('promise should be rejected');
```
Same dealie with `$ERROR`

jugglinmike commented on test/language/statements/async-function/syntax-declaration-no-line-terminator.js:
```diff
@@ -0,0 +1,13 @@
+description: Async function declarations cannot have a line break after `async`
+info: Reference error is thrown due to looking up async in strict mode
+flags: [onlyStrict]
+negative: ReferenceError
```
Since this is a runtime error, could you assert this with `assert.throws`? I still have hopes for gh-196

jugglinmike commented:
These look good, Brian! I had a few ideas for additional tests:

- Assigning to `await` expression or an async function (validating observance
  of the changes to IsValidSimpleAssignmentTarget)
- SyntaxError for "async generators", e.g. `async function* g() {}`
- If I'm reading the spec correctly, then aync functions should not define a
  `prototype` property, so:

      assert.sameValue(async function() {}.prototype, undefined);

ianwjhalliday commented on test/language/expressions/async-arrow-function/early-errors-arrow-arguments-in-formal-parameters.js:
```diff
@@ -0,0 +1,12 @@
+negative: SyntaxError
+---*/
+
+async(arguments) => {  }
```
This isn't in the spec, is it meant to be? I just undid this check for general arrow functions

bterlson replied:
This is true in strict mode. Should be onlyStrict.

ianwjhalliday commented on test/language/expressions/async-arrow-function/early-errors-arrow-eval-in-formal-parameters.js:
```diff
@@ -0,0 +1,11 @@
+negative: SyntaxError
+---*/
+
+async(eval) => {  }
```
Ditto not in spec

bterlson replied:
Ditto.

anba commented on test/language/statements/async-function/evaluation-default-that-throws.js:
```diff
@@ -0,0 +1,19 @@
+  This is different from generators which will throw the error out of the generator
+  when it is called.
+---*/
+async function foo(x = null++) {}
```
`null++` is an early reference error, so the parser will already reject this test file.

bterlson commented:
Additional tests required:

* Mapped vs. unmapped arguments object
* this-value tests
* non-negative super tests

bterlson commented:
Addressed review feedback!

jugglinmike commented on test/language/statements/async-function/evaluation-this-value-global.js:
```diff
@@ -0,0 +1,17 @@
+flags: [noStrict]
+---*/
+
+var glob = Function("return this;")();
```
Could you use the `fnGlobalObject.js` helper here? Using that consistently allows implementers to replace with a host-specific equivalent in cases where the Function constructor is disallowed.

jugglinmike replied:
Actually, scratch that. Since gh-568 and gh-595, we are now encouraging the use of the `this` keyword in the global scope to access the global object.

jugglinmike commented on test/language/statements/async-function/syntax-declaration-no-line-terminator.js:
```diff
@@ -0,0 +1,14 @@
+es7id: pending
+description: Async function declarations cannot have a line break after `async`
+info: Reference error is thrown due to looking up async in strict mode
+flags: [onlyStrict]
```
I believe this test is equally valid for non-strict mode

bterlson replied:
It isn't because async; will just be an identifier lookup that won't throw...

jugglinmike replied:
I think the ReferenceError occurs outside of strict mode

>     ExpressionStatement : Expression ;
>
>     1. Let exprRef be the result of evaluating Expression.
>     2. Return ? GetValue(exprRef). 

and

>     6.2.3.1 GetValue (V)
>
>     1. ReturnIfAbrupt(V).
>     2. If Type(V) is not Reference, return V.
>     3. Let base be GetBase(V).
>     4. If IsUnresolvableReference(V) is true, throw a ReferenceError exception.

which is distinct from

>     PutValue (V, W)
>
>     1. ReturnIfAbrupt(V).
>     2. ReturnIfAbrupt(W).
>     3. If Type(V) is not Reference, throw a ReferenceError exception.
>     4. Let base be GetBase(V).
>     5. If IsUnresolvableReference(V) is true, then
>        a. If IsStrictReference(V) is true, then
>             i. Throw a ReferenceError exception.
>        b. Let globalObj be GetGlobalObject().
>        c. Return ? Set(globalObj, GetReferencedName(V), W, false).

bterlson replied:
Agree, agree. Of course. Thank you!

jugglinmike commented on test/language/statements/class/definition/methods-async-super-call-param.js:
```diff
@@ -0,0 +1,24 @@
+class B extends A {
+  async method(x = super.method()) {
+    var y = await super.method();
+    assert.sameValue(y, 'sup');
```
We should probably include an assertion for the value of `x`. Maybe the default value should be `await super.method()`? Or should that be a separate test?

bterlson replied:
We already have tests for awaiting super in a method, so I just deleted the second await and changed the assertion to x.

jugglinmike commented:
Hi Brian! I've left a couple comments in-line. Some of my feedback doesn't
apply to one specific file, so I'll include it here:

- The latest changeset introduces a few files named `test.txt` that should be
  removed.
- There have been two recent changes to expected test structure that concern
  the files in this patch:
  1. We have moved from `es7id` to the more generic `esid` (see
     https://github.com/tc39/test262/issues/477#issuecomment-178131536 and
     https://github.com/tc39/test262/pull/516)
  2. Asynchronous tests must now specify the `async` tag (see
     https://github.com/tc39/test262/pull/504)

jugglinmike commented:
Another recent change in `master` affects this patch: @michaelficarra's work with Function.prototype.toString. Could you add tests for async functions [following his example](https://github.com/tc39/test262/commit/6b7cbb50350e478a543da40d02492aa8f45a792d)?

leobalter commented:
I just pinged @bterlson today on irc, I hope we can have this merged soon!

bterlson commented:
Sorry missed the notifications on this thread. I will make updates hopefully this week!

bterlson commented:
Addressed all the review feedback here with the exception of the toString tests (which will not run in Chakra due to [other bugs](https://github.com/Microsoft/ChakraCore/issues/971) and so are harder for me to write).

leobalter commented:
LGTM. 

I'm ok to have the toString tests in a follow-up PR.

leobalter commented:
oops, just another detail: `test/language/expressions/async-function/test.txt`

jugglinmike commented:
I believe the ReferenceError in `syntax-declaration-no-line-terminator.js` also occurs outside of strict mode. My rationale is [here](https://github.com/tc39/test262/pull/479#discussion_r62938070).

littledan commented:
These tests are great. @caitp wrote hundreds of procedurally generated syntax tests at https://codereview.chromium.org/1841543003/diff/480001/test/cctest/test-parsing.cc . Any chance at integrating these? Also semantics tests at https://codereview.chromium.org/1895603002 but you might already have equivalent coverage (look at the files starting with test/mjsunit/harmony).

bterlson commented:
I have addressed the remaining feedback (again). Should be good to go.

I did not explore the v8 tests, but I don't think that is on the critical path to landing this PR personally!

littledan commented on test/language/statements/class/definition/methods-async-super-call-param.js:
```diff
@@ -0,0 +1,24 @@
+
+class B extends A {
+  async method(x = super.method()) {
+    assert.sameValue(x, 'sup');
```
Maybe you want 'await x' here.

bterlson replied:
sure do! this test will fail.

bterlson commented:
Fixed. I now think this is ready to go :)

caitp commented on test/language/expressions/object/method-definition/early-errors-object-method-arguments-in-formal-parameters.js:
```diff
@@ -0,0 +1,12 @@
+author: Brian Terlson <brian.terlson@microsoft.com>
+esid: pending
+description: It is a SyntaxError if FormalParameters contains arguments
+negative: SyntaxError
```
This needs to be `onlyStrict` like early-errors-arrow-arguments-in-formal-parameters.js --- this restriction is only enforced in strict mode

caitp commented on test/language/expressions/object/method-definition/early-errors-object-method-eval-in-formal-parameters.js:
```diff
@@ -0,0 +1,12 @@
+author: Brian Terlson <brian.terlson@microsoft.com>
+esid: pending
+description: It is a SyntaxError if FormalParameters contains eval in strict mode
+negative: SyntaxError
```
same thing here --- this restriction is only enforced in strict mode, needs [onlyStrict]

caitp commented on test/language/statements/class/definition/early-errors-class-method-eval-in-formal-parameters.js:
```diff
@@ -0,0 +1,12 @@
+negative: SyntaxError
+---*/
+class Foo {
+  async foo(eval) { }
```
ditto here

caitp commented on test/language/expressions/async-function/early-errors-expression-body-contains-super-call.js:
```diff
@@ -0,0 +1,11 @@
+/*---
+author: Brian Terlson <brian.terlson@microsoft.com>
+esid: pending
+description: It is a syntax error if AsyncFunctionBody contains SuperProperty is true
```
description is misleading, this is a SuperCall

caitp commented on test/language/expressions/async-function/early-errors-expression-formals-contains-super-call.js:
```diff
@@ -0,0 +1,11 @@
+/*---
+author: Brian Terlson <brian.terlson@microsoft.com>
+esid: pending
+description: It is a syntax error if FormalParameters contains SuperProperty is true
```
SuperCall

jugglinmike commented on test/built-ins/Function/prototype/toString/async-method.js:
```diff
@@ -0,0 +1,15 @@
+let g = { /* before */async /* a */ [ /* b */ "g" /* c */ ] /* d */ ( /* e */ ) /* f */ { /* g */ }/* after */ }.g;
+
+assert.sameValue(f.toString(), "async f /* a */ ( /* b */ ) /* c */ { /* d */ }");
+assert.sameValue(g.toString(), "async f /* a */ [ /* b */ \"g\" /* c */ ] /* d */ ( /* e */ ) /* f */ { /* g */ }");
```
This assertion is faulty because "f" is not applicable here. The expected value should be:

    "async /* a */ [ /* b */ \"g\" /* c */ ] /* d */ ( /* e */ ) /* f */ { /* g */ }"

jugglinmike commented:
I've reviewed this patch one more time and found one final detail holding us
back. We've been at this for quite some time, and the fix is straightforward,
so if you'd like, I can fix it as I merge--let me know!

littledan commented:
@jugglinmike Seems like a good idea to me for you to fix as you merge it. Thanks!

leobalter commented:
I'm applying this fix

leobalter commented:
I applied the fix, rebased, squashed and merged this.

:eggplant:
